src-input/duk_bi_encoding.c

```diff
@@ -0,0 +1,348 @@
+
+	nargs = duk_get_top(ctx);
+	if (!duk_is_constructor_call(ctx)) {
+        duk_error(ctx, DUK_ERR_TYPE_ERROR, "constructor must use 'new'");
+    }
+	if (nargs >= 1) {
+		/* XXX: for now, label is ignored */
+		duk_to_string(ctx, 0);
+	}
+	if (nargs >= 2 && !duk_is_null_or_undefined(ctx, 1)) {
+		if (duk_get_prop_string(ctx, 1, "fatal")) {
+			fatal = duk_to_boolean(ctx, -1);
+		}
+		if (duk_get_prop_string(ctx, 1, "ignoreBOM")) {
+			ignore_bom = duk_to_boolean(ctx, -1);
+		}
+	}
+
+	dec_ctx = (duk__decode_context *) duk_alloc(ctx, sizeof(duk__decode_context));
+	dec_ctx->fatal = fatal;
+	dec_ctx->ignore_bom = ignore_bom;
+	duk__utf8_decode_init(dec_ctx);
+
+	duk_push_this(ctx);
+	duk_push_pointer(ctx, (void *) dec_ctx);
+	duk_put_prop_string(ctx, -2, "\xff" "Context");
+	return 0;
+}
+
+DUK_INTERNAL duk_ret_t duk_bi_textdecoder_prototype_encoding_getter(duk_context *ctx) {
+	duk_push_string(ctx, "utf-8");
+	return 1;
+}
+
+DUK_INTERNAL duk_ret_t duk_bi_textdecoder_prototype_fatal_getter(duk_context *ctx) {
+	duk__decode_context *dec_ctx;
+
+	duk_push_this(ctx);
+	duk_get_prop_string(ctx, -1, "\xff" "Context");
+	dec_ctx = (duk__decode_context *) duk_require_pointer(ctx, -1);
+	duk_push_boolean(ctx, dec_ctx->fatal);
+	return 1;
+}
+
+DUK_INTERNAL duk_ret_t duk_bi_textdecoder_prototype_ignorebom_getter(duk_context *ctx) {
+	duk__decode_context *dec_ctx;
+
+	duk_push_this(ctx);
+	duk_get_prop_string(ctx, -1, "\xff" "Context");
+	dec_ctx = (duk__decode_context *) duk_require_pointer(ctx, -1);
+	duk_push_boolean(ctx, dec_ctx->ignore_bom);
+	return 1;
+}
+
+DUK_INTERNAL duk_ret_t duk_bi_textdecoder_prototype_decode(duk_context *ctx) {
+	duk_uint8_t *data = (duk_uint8_t *) "";
+	duk__decode_context *dec_ctx;
+	duk_size_t len = 0;
+	duk_idx_t nargs;
+	duk_bool_t stream = 0;
+	duk_codepoint_t codepoint;
+	int ret;
+	duk_uint8_t *output;
+	duk_uint8_t *in, *out;
+
+	nargs = duk_get_top(ctx);
+	if (nargs >= 1) {
+		data = (duk_uint8_t *) duk_require_buffer_data(ctx, 0, &len);
+	}
+	if (nargs >= 2 && !duk_is_null_or_undefined(ctx, 1)) {
+		if (duk_get_prop_string(ctx, 1, "stream")) {
+			stream = duk_to_boolean(ctx, -1);
+		}
+	}
+	duk_push_this(ctx);
+	duk_get_prop_string(ctx, -1, "\xff" "Context");
+	dec_ctx = (duk__decode_context *) duk_require_pointer(ctx, -1);
+
+	if (len >= (DUK_HBUFFER_MAX_BYTELEN / 3) - 3) {
+		duk_error(ctx, DUK_ERR_TYPE_ERROR, "input data too large");
+	}
+
+	/* XXX: As with encoding, need a better buffer allocation strategy here */
+
+	/* In general, reserving 3x the input length is enough.  In streaming mode,
+	 * if the first byte completes a non-BMP codepoint, it will encode to a
+	 * surrogate pair (6 bytes) so we add an extra 3 bytes to compensate.
+	 */
+	output = (duk_uint8_t *) duk_push_fixed_buffer(ctx, 3 + (3 * len));
+	in = data;
+	out = output;
+	while (in < data + len) {
+		ret = duk__utf8_decode_next(dec_ctx, *in++, &codepoint);
+		if (ret == DUK__U8_CONTINUE) {
+			continue;
+		}
+		if (ret >= DUK__U8_ERROR) {
```

Comment:
Maybe comment that we're handling both `DUK__U8_ERROR` and `DUK__U8_RETRY` here, or maybe even assert:

    DUK_ASSERT(DUK__U8_RETRY > DUK__U8_ERROR);
    DUK_ASSERT(ret == DUK__U8_ERROR || ret == DUK__U8_RETRY);